runtime/runtime/src/balance_checker.rs

```diff
@@ -589,4 +604,145 @@ mod tests {
+                    .insert(0, TrieQueueIndices { first_index: 0, next_available_index: 2 });
+
+                set(trie_update, TrieKey::BufferedReceiptIndices, &indices);
+                set(
+                    trie_update,
+                    TrieKey::BufferedReceipt { receiving_shard: 0, index: 1 },
+                    &new_receipt,
+                );
+            },
+        );
+
+        check_balance(
+            &RuntimeConfig::test(),
+            &final_state,
+            &None,
+            &[],
+            &[],
+            &[tx],
+            &[],
+            &ApplyStats {
+                // send gas was burnt on this shard, exec gas is part of the receipt value
+                tx_burnt_amount: send_gas as Balance * gas_price,
+                gas_deficit_amount: 0,
+                other_burnt_amount: 0,
+                slashed_burnt_amount: 0,
+            },
+        )
+        .unwrap();
+    }
+
+    /// When forwarding a receipt from the outgoing buffer, its balance must
+    /// must be picked up by the balance checker. Test it by removing a transfer
+    /// receipt from the buffer and adding it to the outgoing receipts.
+    #[test]
+    fn test_remove_buffered_receipt_balance() {
+        let account_id = alice_account();
+        let deposit = 100;
+        let gas_price = 10;
+
+        let tx = transfer_tx(account_id, bob_account(), deposit);
+        let receipt0 = extract_transfer_receipt(&tx, gas_price, deposit);
+        let receipt1 = extract_transfer_receipt(&tx, gas_price, deposit + 5);
+
+        let final_state = prepare_state_change(
+            |trie_update| {
+                // store 2 receipts with balance in the receipt buffer
+                let mut indices = BufferedReceiptIndices::default();
+                indices
+                    .shard_buffers
+                    .insert(0, TrieQueueIndices { first_index: 0, next_available_index: 2 });
+
+                set(trie_update, TrieKey::BufferedReceiptIndices, &indices);
+                set(
+                    trie_update,
+                    TrieKey::BufferedReceipt { receiving_shard: 0, index: 0 },
+                    &receipt0,
+                );
+                set(
+                    trie_update,
+                    TrieKey::BufferedReceipt { receiving_shard: 0, index: 1 },
+                    &receipt1,
+                );
+            },
+            |trie_update| {
+                // remove 1 receipt at index 0
+                let mut indices = BufferedReceiptIndices::default();
+                indices
+                    .shard_buffers
+                    .insert(0, TrieQueueIndices { first_index: 1, next_available_index: 2 });
+
+                set(trie_update, TrieKey::BufferedReceiptIndices, &indices);
+                trie_update.remove(TrieKey::BufferedReceipt { receiving_shard: 0, index: 0 });
+            },
+        );
+
+        let outgoing_receipts = [receipt0];
+        check_balance(
```

Comment:
You can also consider adding a "negative" test where it's supposed to fail and return the right error. Totally optional.

Reply:
Good idea, let me add that quickly.

Reply:
Added now, will merge once CI passes